packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -2452,6 +2521,74 @@ export class SnapController extends BaseController<
     }
   }
 
+  #getRollbackSnapshot(snapId: SnapId): RollbackSnapshot | undefined {
+    return this._rollbackSnapshots.get(snapId);
+  }
+
+  #createRollbackSnapshot(snapId: SnapId): RollbackSnapshot {
+    assert(
+      this.#getRollbackSnapshot(snapId) === undefined,
+      new Error(`Snap "${snapId}" rollback snapshot already exists.`),
+    );
+
+    this._rollbackSnapshots.set(snapId, {
+      statePatches: [],
+      sourceCode: '',
+      permissions: { revoked: null, granted: [], requestData: null },
+      newVersion: '',
+    });
+    return this.#getRollbackSnapshot(snapId) as RollbackSnapshot;
+  }
+
+  async #rollbackSnap(snapId: SnapId) {
+    const rollbackSnapshot = this.#getRollbackSnapshot(snapId);
+    if (!rollbackSnapshot) {
+      throw new Error('A snapshot does not exist for this snap.');
+    }
+
+    const { statePatches, sourceCode, permissions } = rollbackSnapshot;
+
+    if (statePatches?.length) {
+      this.applyPatches(statePatches);
+    }
+
+    if (sourceCode) {
+      const runtime = this.#getRuntimeExpect(snapId);
+      runtime.sourceCode = sourceCode;
+    }
+
+    if (permissions.revoked && Object.keys(permissions.revoked).length) {
+      this.messagingSystem.call('PermissionController:grantPermissions', {
+        approvedPermissions: permissions.revoked as RequestedPermissions,
+        subject: { origin: snapId },
+        requestData: permissions.requestData as Record<string, unknown>,
+      });
+    }
+
+    if (permissions.granted?.length) {
+      this.messagingSystem.call('PermissionController:revokePermissions', {
+        [snapId]: permissions.granted as NonEmptyArray<string>,
+      });
+    }
+
+    await this.stopSnap(snapId, SnapStatusEvents.Stop);
+
+    const truncatedSnap = this.getTruncatedExpect(snapId);
+
+    this.messagingSystem.publish(
+      'SnapController:snapRolledback',
+      truncatedSnap,
+      rollbackSnapshot.newVersion,
+    );
+  }
+
+  async #rollbackSnaps(snapIds: SnapId[]) {
+    for (const snapId of snapIds) {
+      await this.#rollbackSnap(snapId);
+      this._rollbackSnapshots.delete(snapId);
```

Comment:
Perhaps `rollbackSnap` should do this so we don't forget about it in the future if we call that directly?